Allow to override the default theme

skjnldsv opened the pull request:
Still allow fonts changing
![Capture d’écran_2022-04-29_11-51-17](https://user-images.githubusercontent.com/14975046/165923237-7e7600b3-09ed-432b-9eec-7544424e477b.png)

skjnldsv committed:
Allow to override the default theme

Signed-off-by: John Molakvoæ <skjnldsv@protonmail.com>

artonge commented on config/config.sample.php:
```diff
@@ -1789,6 +1789,12 @@
+ * Enforce the user theme. This will disable the user theming settings
+ * This must be a valid ITheme ID
+ */
+'default_theme' => '',
```
`default_theme` sounds like it can be changed be the user.

```suggestion
'enforce_theme' => '',
```

skjnldsv replied:
Done

szaimen commented:
Do we wanna add a test for this?

skjnldsv commented:
> Should we add a test for this?

Done

szaimen commented on config/config.sample.php:
```diff
@@ -1789,6 +1789,12 @@
+ * Enforce the user theme. This will disable the user theming settings
+ * This must be a valid ITheme ID
+ */
+'enforce_theme' => '',
```
maybe we should document available theme ids here?

szaimen replied:
ping @skjnldsv

skjnldsv replied:
Not sure tbh, most customers wanting to enforce this would most ikely have custom themes, so we don't know the ID

szaimen replied:
but we could at least add the default ones?

skjnldsv replied:
Done

szaimen replied:
thanks!

szaimen commented:
Seems to work but after removing 'enforce_theme' again I run into an internal server issue:
```
{
  "reqId": "3JUEJbEbfDP4SQCUtOyh",
  "level": 3,
  "time": "2022-04-29T18:06:26+00:00",
  "remoteAddr": "192.168.146.1",
  "user": "admin",
  "app": "PHP",
  "method": "GET",
  "url": "/index.php/apps/dashboard/",
  "message": "array_merge(): Expected parameter 2 to be an array, object given at /var/www/nextcloud/apps/theming/lib/Service/ThemesService.php#161",
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.41 Safari/537.36",
  "version": "25.0.0.1",
  "exception": {
    "Exception": "Error",
    "Message": "array_merge(): Expected parameter 2 to be an array, object given at /var/www/nextcloud/apps/theming/lib/Service/ThemesService.php#161",
    "Code": 0,
    "Trace": [
      {
        "function": "onError",
        "class": "OC\\Log\\ErrorHandler",
        "type": "::",
        "args": [
          2,
          "array_merge(): Expected parameter 2 to be an array, object given",
          "/var/www/nextcloud/apps/theming/lib/Service/ThemesService.php",
          161,
          {
            "user": {
              "__class__": "OC\\User\\User"
            },
            "enforcedTheme": "dark-highcontrast",
            "enabledThemes": {
              "__class__": "stdClass",
              "0": "dark-highcontrast",
              "1": "default"
            }
          }
        ]
      },
      {
        "file": "/var/www/nextcloud/apps/theming/lib/Service/ThemesService.php",
        "line": 161,
        "function": "array_merge",
        "args": [
          [
            "dark-highcontrast"
          ],
          {
            "__class__": "stdClass",
            "0": "dark-highcontrast",
            "1": "default"
          }
        ]
      },
      {
        "file": "/var/www/nextcloud/apps/theming/lib/Service/JSDataService.php",
        "line": 62,
        "function": "getEnabledThemes",
        "class": "OCA\\Theming\\Service\\ThemesService",
        "type": "->",
        "args": []
      },
      {
        "function": "jsonSerialize",
        "class": "OCA\\Theming\\Service\\JSDataService",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/nextcloud/lib/private/InitialStateService.php",
        "line": 68,
        "function": "json_encode",
        "args": [
          {
            "__class__": "OCA\\Theming\\Service\\JSDataService"
          }
        ]
      },
      {
        "file": "/var/www/nextcloud/lib/private/InitialStateService.php",
        "line": 89,
        "function": "provideInitialState",
        "class": "OC\\InitialStateService",
        "type": "->",
        "args": [
          "theming",
          "data",
          {
            "__class__": "OCA\\Theming\\Service\\JSDataService"
          }
        ]
      },
      {
        "file": "/var/www/nextcloud/lib/private/InitialStateService.php",
        "line": 140,
        "function": "invokeLazyStateCallbacks",
        "class": "OC\\InitialStateService",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/nextcloud/lib/private/TemplateLayout.php",
        "line": 284,
        "function": "getInitialStates",
        "class": "OC\\InitialStateService",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/nextcloud/lib/private/legacy/OC_Template.php",
        "line": 185,
        "function": "__construct",
        "class": "OC\\TemplateLayout",
        "type": "->",
        "args": [
          "error",
          ""
        ]
      },
      {
        "file": "/var/www/nextcloud/lib/private/Template/Base.php",
        "line": 132,
        "function": "fetchPage",
        "class": "OC_Template",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/nextcloud/lib/private/legacy/OC_Template.php",
        "line": 334,
        "function": "printPage",
        "class": "OC\\Template\\Base",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/nextcloud/index.php",
        "line": 88,
        "function": "printExceptionErrorPage",
        "class": "OC_Template",
        "type": "::",
        "args": [
          {
            "__class__": "TypeError"
          },
          500
        ]
      }
    ],
    "File": "/var/www/nextcloud/lib/private/Log/ErrorHandler.php",
    "Line": 92,
    "CustomMessage": "--"
  }
}
{
  "reqId": "3JUEJbEbfDP4SQCUtOyh",
  "level": 3,
  "time": "2022-04-29T18:06:26+00:00",
  "remoteAddr": "192.168.146.1",
  "user": "admin",
  "app": "PHP",
  "method": "GET",
  "url": "/index.php/apps/dashboard/",
  "message": "TypeError: Return value of OCA\\Theming\\Service\\ThemesService::getEnabledThemes() must be of the type array, null returned at /var/www/nextcloud/apps/theming/lib/Service/ThemesService.php#161",
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.41 Safari/537.36",
  "version": "25.0.0.1"
}
```

skjnldsv committed:
Fix theming setEnabledThemes unique array

Signed-off-by: John Molakvoæ <skjnldsv@protonmail.com>

skjnldsv commented:
> Seems to work but after removing 'enforce_theme' again I run into an internal server issue:
> 
> click to expand

I had that a few times, couldn't figure out why.
Added some tests and fixed it . It was an issue with the order or `array_values` and `array_unique` :)
Thanks for the help Simon!

```php
1) OCA\Theming\Tests\Service\ThemesServiceTest::testSetEnabledThemes with data set #2 (array('dark', 'dark', 'opendyslexic'), array('dark', 'opendyslexic'))
Expectation failed for method name is "setUserValue" when invoked 1 time(s)
Parameter 3 for invocation OCP\IConfig::setUserValue('user', 'theming', 'enabled-themes', '{"0":"dark","2":"opendyslexic"}', null) does not match expected value.
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'["dark","opendyslexic"]'
+'{"0":"dark","2":"opendyslexic"}'

/var/www/nextcloud/apps/theming/lib/Service/ThemesService.php:179
/var/www/nextcloud/tests/lib/TestCase.php:226
/var/www/nextcloud/apps/theming/tests/Service/ThemesServiceTest.php:270

```

skjnldsv committed:
Add enforce_theme example IDs

Signed-off-by: John Molakvoæ <skjnldsv@protonmail.com>

szaimen reviewed (approved):
Having a not-valid id now will remove all themes from the page which might lead to some users reporting a bug.

Otherwiese works but didn't review the code.

skjnldsv commented:
/rebase
